synapse/handlers/room_summary.py

```diff
@@ -720,7 +712,7 @@ async def _build_room_entry(self, room_id: str, for_federation: bool) -> JsonDic
                 stats["history_visibility"] == HistoryVisibility.WORLD_READABLE
             ),
             "guest_can_join": stats["guest_access"] == "can_join",
-            "room_type": create_event.content.get(EventContentFields.ROOM_TYPE),
+            "room_type": stats["room_type"],
```

Comment:
(Somewhat random spot to start a thread on.)

This is failing complement tests, I'm not really sure *why*, but the `room_type` is coming back as `None`. As far as I can tell the create event has been proper "processed" though (which is why it exists at all?)

I'm not really sure what's happening, but adding any logging to the stats processing seems to fix the tests.

----

Anyone have thoughts of what could be going on or how to do debug this?

Reply:
As a note: `get_room_with_stats_txn` looks like
```
                SELECT ... FROM rooms
                LEFT JOIN room_stats_state state USING (room_id)
                LEFT JOIN room_stats_current curr USING (room_id)
```
so the `room_stats_state` row may not necessarily exist by the time this request is made — it's updated in the background, after all. In this case, it indeed will appear as `None`.
(Though in my testing there were also `"joined_members": 1` which suggests the `m.room.member` rows have been processed.)

As you noted, adding any actual logging calls seems to make the flake go away, so I resorted to adding some poor man's logging calls that just append text into a buffer and then I truly log it out at the time of the request. I added such calls into some places in the delta/incremental updater for room stats.

<details>
<summary>crude logs</summary>

Successful run
```
synapse_main | 2022-11-07 18:42:28,131 - synapse.handlers.room_summary - 704 - ERROR - GET-17 - XXX room stats {'room_id': '!OpznUsgeOfivwuJKib:hs1', 'name': None, 'canonical_alias': None, 'joined_members': 1, 'joined_local_members': 1, 'version': '9', 'creator': '@alice:hs1', 'encryption': None, 'federatable': True, 'public': False, 'join_rules': 'public', 'guest_access': None, 'history_visibility': 'world_readable', 'state_events': 8, 'avatar': None, 'topic': None, 'room_type': 'm.space'}
synapse_main | 2022-11-07 18:42:28,131 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  10 = 'm.room.create'.''
synapse_main | 2022-11-07 18:42:28,131 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  11 = 'm.room.member'.'@alice:hs1'
synapse_main | 2022-11-07 18:42:28,131 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  12 = 'm.room.power_levels'.''
synapse_main | 2022-11-07 18:42:28,131 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  13 = 'm.room.join_rules'.''
synapse_main | 2022-11-07 18:42:28,131 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  14 = 'm.room.history_visibility'.''
synapse_main | 2022-11-07 18:42:28,131 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  30 = 'm.space.child'.'!rvrMiUcNFYCFDhoJes:hs1'
synapse_main | 2022-11-07 18:42:28,131 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  31 = 'm.space.child'.'!GgEmifmSpUQvTQTCqW:hs1'
synapse_main | 2022-11-07 18:42:28,131 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  32 = 'm.space.child'.'!rkxurPkTXoznZJhjPU:hs2'
synapse_main | 2022-11-07 18:42:28,131 - synapse.handlers.room_summary - 710 - ERROR - GET-17 - XXX lbuf ---
synapse_main | Got pos 9 from stats_incremental_position
synapse_main | 1667846546.8114002 Processing room stats 9->10
synapse_main | Handling 1 state deltas
synapse_main | Initial token for room !OpznUsgeOfivwuJKib:hs1: None
synapse_main | 1667846546.8187804 (B) Proc ev typ = 'm.room.create'
synapse_main | 1667846546.8187878 (B)Setting room type of !OpznUsgeOfivwuJKib:hs1 to m.space
synapse_main | 1667846546.8187973 Upserting rss !OpznUsgeOfivwuJKib:hs1 {'is_federatable': True, 'room_type': 'm.space'}
synapse_main | 1667846546.8224208 Upserted rss !OpznUsgeOfivwuJKib:hs1
synapse_main | 1667846546.831069 Handled room stats to 9 -> 10
synapse_main | 1667846546.8458247 Processing room stats 10->11
synapse_main | Handling 1 state deltas
synapse_main | Initial token for room !OpznUsgeOfivwuJKib:hs1: None
synapse_main | 1667846546.8485677 (B) Proc ev typ = 'm.room.member'
synapse_main | 1667846546.848585 Upserting rss !OpznUsgeOfivwuJKib:hs1 {}
synapse_main | 1667846546.8518713 Upserted rss !OpznUsgeOfivwuJKib:hs1
synapse_main | 1667846546.8590639 Handled room stats to 10 -> 11
synapse_main | 1667846546.8803842 Processing room stats 11->12
synapse_main | Handling 1 state deltas
synapse_main | Initial token for room !OpznUsgeOfivwuJKib:hs1: None
synapse_main | 1667846546.8892975 (B) Proc ev typ = 'm.room.power_levels'
synapse_main | 1667846546.889311 Upserting rss !OpznUsgeOfivwuJKib:hs1 {}
synapse_main | 1667846546.8938875 Upserted rss !OpznUsgeOfivwuJKib:hs1
synapse_main | 1667846546.8984704 Handled room stats to 11 -> 12
synapse_main | 1667846546.9118683 Processing room stats 12->14
synapse_main | Handling 2 state deltas
synapse_main | Initial token for room !OpznUsgeOfivwuJKib:hs1: None
synapse_main | 1667846546.9156551 (B) Proc ev typ = 'm.room.history_visibility'
..
```


Failed run
```
synapse_main | 2022-11-07 18:35:54,149 - synapse.handlers.room_summary - 704 - ERROR - GET-17 - XXX room stats {'room_id': '!ikIhUGvNXNtUEfqdkw:hs1', 'name': None, 'canonical_alias': None, 'joined_members': 1, 'joined_local_members': 1, 'version': '9', 'creator': '@alice:hs1', 'encryption': None, 'federatable': False, 'public': False, 'join_rules': 'public', 'guest_access': None, 'history_visibility': 'world_readable', 'state_events': 7, 'avatar': None, 'topic': None, 'room_type': None}
synapse_main | 2022-11-07 18:35:54,150 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO   9 = 'm.room.create'.''
synapse_main | 2022-11-07 18:35:54,150 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  10 = 'm.room.member'.'@alice:hs1'
synapse_main | 2022-11-07 18:35:54,150 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  11 = 'm.room.power_levels'.''
synapse_main | 2022-11-07 18:35:54,150 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  12 = 'm.room.join_rules'.''
synapse_main | 2022-11-07 18:35:54,150 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  13 = 'm.room.history_visibility'.''
synapse_main | 2022-11-07 18:35:54,150 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  29 = 'm.space.child'.'!ViAGENnBLueltrGwuA:hs1'
synapse_main | 2022-11-07 18:35:54,150 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  30 = 'm.space.child'.'!obSXlQzUdyFOHjvrqC:hs1'
synapse_main | 2022-11-07 18:35:54,150 - synapse.handlers.room_summary - 708 - ERROR - GET-17 - XXX SO  31 = 'm.space.child'.'!xgEEMEfreFlEVBuZKs:hs2'
synapse_main | 2022-11-07 18:35:54,150 - synapse.handlers.room_summary - 710 - ERROR - GET-17 - XXX lbuf ---
synapse_main | Got pos 8 from stats_incremental_position
synapse_main | 1667846152.78026 Processing room stats 8->9
synapse_main | Handling 0 state deltas
synapse_main | 1667846152.7870986 Handled room stats to 8 -> 9
synapse_main | 1667846152.8076231 Processing room stats 9->10
synapse_main | Handling 1 state deltas
synapse_main | Initial token for room !ikIhUGvNXNtUEfqdkw:hs1: None
synapse_main | 1667846152.8114643 (B) Proc ev typ = 'm.room.member'
synapse_main | 1667846152.8114808 Upserting rss !ikIhUGvNXNtUEfqdkw:hs1 {}
synapse_main | 1667846152.8168423 Upserted rss !ikIhUGvNXNtUEfqdkw:hs1
synapse_main | 1667846152.827462 Handled room stats to 9 -> 10
synapse_main | 1667846152.8496284 Processing room stats 10->11
synapse_main | Handling 1 state deltas
synapse_main | Initial token for room !ikIhUGvNXNtUEfqdkw:hs1: None
synapse_main | 1667846152.8593013 (B) Proc ev typ = 'm.room.power_levels'
synapse_main | 1667846152.8593154 Upserting rss !ikIhUGvNXNtUEfqdkw:hs1 {}
synapse_main | 1667846152.8636587 Upserted rss !ikIhUGvNXNtUEfqdkw:hs1
synapse_main | 1667846152.8682039 Handled room stats to 10 -> 11
synapse_main | 1667846152.882483 Processing room stats 11->13
synapse_main | Handling 2 state deltas
synapse_main | Initial token for room !ikIhUGvNXNtUEfqdkw:hs1: None
synapse_main | 1667846152.8867443 (B) Proc ev typ = 'm.room.history_visibility'
synapse_main | Initial token for room !ikIhUGvNXNtUEfqdkw:hs1: None
synapse_main | 1667846152.8885934 (B) Proc ev typ = 'm.room.join_rules'
synapse_main | 1667846152.8886034 Upserting rss !ikIhUGvNXNtUEfqdkw:hs1 {'history_visibility': 'world_readable', 'join_rules': 'public'}
synapse_main | 1667846152.8965864 Upserted rss !ikIhUGvNXNtUEfqdkw:hs1
synapse_main | 1667846152.9042344 Handled room stats to 11 -> 13
synapse_main | 1667846152.9462326 Processing room stats 13->14
synapse_main | Handling 1 state deltas
..
```

</details>

branch with poor man's logging: `rei/clokep/spaces-room-type`

What I find notable: in the successful case, there are log entries corresponding to the `m.room.create` event, but in the failed case, it (usually(?)) skips to `m.room.member`.
This is despite the stream position moving in a way that looks like it should include the `m.room.create` event (??)
I'm not sure why the stream positions seem to be consistently off by 1 in the successful case, either. Weird. Possibly something there.

Need to hop off for tonight though, so didn't get any further, but it looks like a race where the right state deltas aren't being returned in time?

Reply:
Oh, that's a good idea of appending to a buffer instead. Thanks! I'll try to grok what's going on with the streams.

Reply:
Using some of @reivilibre's debugging (and adding some more) what I'm seeing is:

1. State deltas are queried for the stream position covering the create event.
2. For some reason, when this occurs there are no state deltas available (`current_state_delta_stream` is empty for the queried stream IDs), this results in the `current_state_stats` table not being updated.
3. The second event in the room (`m.room.member`) *is* available by the time the state deltas for it are queried.
4. At some point after the above, the `current_state_delta_stream` is properly updated to insert the create event at the proper stream ID.

So I think there's a race somewhere?

As far as I can tell this is *not* related to the recent partial state work *or* the recent batching of room creation events, but I'm not 100% sure on this.

Reply:
Tracking this down a bit more, it seems that the call to `StreamChangeCache.entity_has_changed` is being ignored since the stream position of the `m.room.create` event is equal to the earliest known stream position (which gets set due to: https://github.com/matrix-org/synapse/blob/c3a4780080a5bcb04132283c0f32f7452655792a/synapse/storage/database.py#L2295-L2297)

Reply:
I'm hoping this was fixed by #14435.

Reply:
I think this is *actually* fixed by #14639.